File: src/typescript/src/cdk/base-provider.ts
/**
 * Base provider connector implementation.
 *
 * Implements a full provider interface with OpenAI-compatible default behavior.
 * Subclasses override protected hook methods to adapt to non-OpenAI APIs
 * without reimplementing transport, retries, or error classification.
 *
 * HTTP transport uses Node.js built-in http/https modules so the package
 * has zero external dependencies.
 */

import * as http from 'http';
import * as https from 'https';
import { URL } from 'url';

import { RuntimeEnvironment } from '../interfaces/runtime';
import {
  ChatMessage,
  CompletionChoice,
  CompletionRequest,
  CompletionResponse,
  ErrorClassification,
  ModelInfo,
  ModelPricing,
  QuotaStatus,
  RateLimitStatus,
  TokenUsage,
  createDefaultCompletionResponse as createCompletionResponse,
  createDefaultTokenUsage as createTokenUsage,
} from '../interfaces/provider';

export interface BaseProviderConfig {
  baseUrl: string;
  apiKey: string;
  models: ModelInfo[];
  timeout: number;
  maxRetries: number;
  authMethod: string;
  retryableCodes: number[];
  nonRetryableCodes: number[];
  capabilities: string[];
}

export function createBaseProviderConfig(
  partial?: Partial<BaseProviderConfig>
): BaseProviderConfig {
  return {
    baseUrl: '',
    apiKey: '',
    models: [],
    timeout: 30,
    maxRetries: 3,
    authMethod: 'api_key',
    retryableCodes: [429, 500, 502, 503],
    nonRetryableCodes: [400, 401, 403],
    capabilities: ['generation.text-generation.chat-completion'],
    ...partial,
  };
}

/**
 * Error class for HTTP errors with a status code.
 */
export class HttpError extends Error {
  public readonly statusCode: number;
  public readonly responseBody: string;
  public readonly headers: Record<string, string>;

  constructor(statusCode: number, responseBody: string, headers: Record<string, string> = {}) {
    super(`HTTP ${statusCode}: ${responseBody.substring(0, 200)}`);
    this.name = 'HttpError';
    this.statusCode = statusCode;
    this.responseBody = responseBody;
    this.headers = headers;
  }
}

/**
 * Base implementation of the ProviderConnector interface.
 *
 * Provides an OpenAI-compatible default behavior for all methods.
 * Subclasses override protected hook methods to adapt to non-OpenAI
 * APIs without reimplementing transport, retries, or error handling.
 */
export class BaseProvider {
  static readonly RUNTIME: RuntimeEnvironment = RuntimeEnvironment.NODE_ONLY;

  protected _config: BaseProviderConfig;
  protected _requestCount: number = 0;
  protected _tokensUsed: number = 0;
  protected _modelsByid: Map<string, ModelInfo>;

  constructor(config: BaseProviderConfig) {
    this._config = config;
    this._modelsByid = new Map();
    for (const m of config.models) {
      this._modelsByid.set(m.id, m);
    }
  }

  // -- Model Execution -------------------------------------------------------

  async complete(request: CompletionRequest): Promise<CompletionResponse> {
    const payload = this._buildRequestPayload(request);
    const headers = this._buildHeaders();
    const endpoint = this._getCompletionEndpoint();

    let lastError: Error | null = null;
    for (let attempt = 0; attempt <= this._config.maxRetries; attempt++) {
      try {
        const data = await this._httpPost(endpoint, payload, headers);
        const result = this._parseResponse(data);
        this.reportUsage(request.model, result.usage);
        return result;
      } catch (exc: any) {
        lastError = exc;
        const classification = this.classifyError(exc);
        if (!classification.retryable || attempt === this._config.maxRetries) {
          throw exc;
        }
        let retryAfter: number = Math.pow(2, attempt);
        if (exc instanceof HttpError && exc.headers['retry-after']) {
          const parsed = parseFloat(exc.headers['retry-after']);
          if (!isNaN(parsed)) {
            retryAfter = parsed;
          }
        }
        await this._sleep(retryAfter * 1000);
      }
    }
    throw lastError!;
  }

  async *stream(request: CompletionRequest): AsyncIterableIterator<CompletionResponse> {
    const payload = this._buildRequestPayload(request);
    payload.stream = true;
    const headers = this._buildHeaders();
    const endpoint = this._getCompletionEndpoint();

    const lines = await this._httpPostStream(endpoint, payload, headers);
    for (const line of lines) {
      if (!line || !line.startsWith('data: ')) {
        continue;
      }
      const dataStr = line.substring(6);
      if (dataStr.trim() === '[DONE]') {
        break;
      }
      const chunk = this._parseSseChunk(dataStr);
      if (chunk !== null) {
        yield chunk;
      }
    }
  }

  // -- Capabilities ----------------------------------------------------------

  getCapabilities(): string[] {
    return [...this._config.capabilities];
  }

  supports(capability: string): boolean {
    return this._config.capabilities.includes(capability);
  }

  // -- Model Catalogue -------------------------------------------------------

  listModels(): ModelInfo[] {
    return [...this._config.models];
  }

  getModelInfo(modelId: string): ModelInfo {
    const info = this._modelsByid.get(modelId);
    if (!info) {
      throw new Error(`Model not found: ${modelId}`);
    }
    return info;
  }

  // -- Quota & Rate Limits ---------------------------------------------------

  checkQuota(): QuotaStatus {
    return { used: this._requestCount };
  }

  getRateLimits(): RateLimitStatus {
    return {};
  }

  // -- Cost & Pricing --------------------------------------------------------

  getPricing(modelId: string): ModelPricing {
    const info = this.getModelInfo(modelId);
    if (!info.pricing) {
      throw new Error(`No pricing configured for model: ${modelId}`);
    }
    return info.pricing;
  }

  reportUsage(modelId: string, usage: TokenUsage): void {
    this._requestCount += 1;
    this._tokensUsed += usage.totalTokens;
  }

  // -- Error Classification --------------------------------------------------

  classifyError(error: Error): ErrorClassification {
    let statusCode: number | undefined;
    if (error instanceof HttpError) {
      statusCode = error.statusCode;
    }

    if (statusCode === undefined) {
      return { retryable: false, category: 'unknown', message: error.message };
    }

    if (this._config.retryableCodes.includes(statusCode)) {
      return {
        retryable: true,
        errorCode: statusCode,
        category: statusCode === 429 ? 'rate_limit' : 'server',
        message: error.message,
      };
    }

    if (this._config.nonRetryableCodes.includes(statusCode)) {
      const category = statusCode === 401 || statusCode === 403 ? 'auth' : 'client';
      return {
        retryable: false,
        errorCode: statusCode,
        category,
        message: error.message,
      };
    }

    return {
      retryable: false,
      errorCode: statusCode,
      category: 'unknown',
      message: error.message,
    };
  }

  isRetryable(error: Error): boolean {
    return this.classifyError(error).retryable;
  }

  // -- Protected Hooks -------------------------------------------------------

  protected _buildRequestPayload(request: CompletionRequest): Record<string, any> {
    const payload: Record<string, any> = {
      model: request.model,
      messages: request.messages,
    };
    if (request.temperature !== undefined) {
      payload.temperature = request.temperature;
    }
    if (request.maxTokens != null) {
      payload.max_tokens = request.maxTokens;
    }
    if (request.tools && request.tools.length > 0) {
      payload.tools = request.tools;
    }
    if (request.stream) {
      payload.stream = true;
    }
    return payload;
  }

  protected _parseResponse(data: Record<string, any>): CompletionResponse {
    const usageData = data.usage || {};
    const rawChoices = data.choices || [];
    const choices: CompletionChoice[] = [];
    for (const raw of rawChoices) {
      const msg = raw.message;
      choices.push({
        index: raw.index ?? 0,
        message: msg
          ? { role: msg.role || 'assistant', content: msg.content ?? undefined }
          : undefined,
        finishReason: raw.finish_reason ?? undefined,
      });
    }
    return createCompletionResponse({
      id: data.id || '',
      model: data.model || '',
      choices,
      usage: createTokenUsage({
        promptTokens: usageData.prompt_tokens || 0,
        completionTokens: usageData.completion_tokens || 0,
        totalTokens: usageData.total_tokens || 0,
      }),
    });
  }

  protected _buildHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    if (this._config.apiKey) {
      headers['Authorization'] = `Bearer ${this._config.apiKey}`;
    }
    return headers;
  }

  protected _getCompletionEndpoint(): string {
    const base = this._config.baseUrl.replace(/\/+$/, '');
    return `${base}/v1/chat/completions`;
  }

  protected _parseSseChunk(line: string): CompletionResponse | null {
    let data: Record<string, any>;
    try {
      data = JSON.parse(line);
    } catch {
      return null;
    }
    const rawChoices = data.choices || [];
    if (rawChoices.length === 0) {
      return null;
    }
    const choices: CompletionChoice[] = [];
    for (const raw of rawChoices) {
      const delta = raw.delta;
      choices.push({
        index: raw.index ?? 0,
        delta: delta
          ? { role: delta.role || 'assistant', content: delta.content ?? undefined }
          : undefined,
        finishReason: raw.finish_reason ?? undefined,
      });
    }
    return createCompletionResponse({
      id: data.id || '',
      model: data.model || '',
      choices,
      usage: createTokenUsage(),
    });
  }

  async close(): Promise<void> {
    // No-op by default
  }

  // -- HTTP Transport (Node.js built-in) -------------------------------------

  protected _httpPost(
    url: string,
    payload: Record<string, any>,
    headers: Record<string, string>
  ): Promise<Record<string, any>> {
    return new Promise((resolve, reject) => {
      const body = JSON.stringify(payload);
      const parsedUrl = new URL(url);
      const transport = parsedUrl.protocol === 'https:' ? https : http;

      const options: http.RequestOptions = {
        method: 'POST',
        hostname: parsedUrl.hostname,
        port: parsedUrl.port || (parsedUrl.protocol === 'https:' ? 443 : 80),
        path: parsedUrl.pathname + parsedUrl.search,
        headers: {
          ...headers,
          'Content-Length': Buffer.byteLength(body),
        },
        timeout: this._config.timeout * 1000,
      };

      const req = transport.request(options, (res) => {
        const chunks: Buffer[] = [];
        res.on('data', (chunk: Buffer) => chunks.push(chunk));
        res.on('end', () => {
          const responseBody = Buffer.concat(chunks).toString('utf-8');
          const statusCode = res.statusCode || 0;
          if (statusCode >= 200 && statusCode < 300) {
            try {
              resolve(JSON.parse(responseBody));
            } catch {
              reject(new Error(`Invalid JSON response: ${responseBody.substring(0, 200)}`));
            }
          } else {
            const responseHeaders: Record<string, string> = {};
            for (const [key, value] of Object.entries(res.headers)) {
              if (typeof value === 'string') {
                responseHeaders[key.toLowerCase()] = value;
              }
            }
            reject(new HttpError(statusCode, responseBody, responseHeaders));
          }
        });
      });

      req.on('error', (err) => reject(err));
      req.on('timeout', () => {
        req.destroy();
        reject(new Error(`Request timed out after ${this._config.timeout}s`));
      });

      req.write(body);
      req.end();
    });
  }

  protected _httpPostStream(
    url: string,
    payload: Record<string, any>,
    headers: Record<string, string>
  ): Promise<string[]> {
    return new Promise((resolve, reject) => {
      const body = JSON.stringify(payload);
      const parsedUrl = new URL(url);
      const transport = parsedUrl.protocol === 'https:' ? https : http;

      const options: http.RequestOptions = {
        method: 'POST',
        hostname: parsedUrl.hostname,
        port: parsedUrl.port || (parsedUrl.protocol === 'https:' ? 443 : 80),
        path: parsedUrl.pathname + parsedUrl.search,
        headers: {
          ...headers,
          'Content-Length': Buffer.byteLength(body),
        },
        timeout: this._config.timeout * 1000,
      };

      const req = transport.request(options, (res) => {
        const chunks: Buffer[] = [];
        res.on('data', (chunk: Buffer) => chunks.push(chunk));
        res.on('end', () => {
          const responseBody = Buffer.concat(chunks).toString('utf-8');
          const statusCode = res.statusCode || 0;
          if (statusCode >= 200 && statusCode < 300) {
            resolve(responseBody.split(/\r?\n/));
          } else {
            const responseHeaders: Record<string, string> = {};
            for (const [key, value] of Object.entries(res.headers)) {
              if (typeof value === 'string') {
                responseHeaders[key.toLowerCase()] = value;
              }
            }
            reject(new HttpError(statusCode, responseBody, responseHeaders));
          }
        });
      });

      req.on('error', (err) => reject(err));
      req.on('timeout', () => {
        req.destroy();
        reject(new Error(`Request timed out after ${this._config.timeout}s`));
      });

      req.write(body);
      req.end();
    });
  }

  protected _httpPostRaw(
    url: string,
    payload: Record<string, any>,
    headers: Record<string, string>
  ): Promise<Buffer> {
    return new Promise((resolve, reject) => {
      const body = JSON.stringify(payload);
      const parsedUrl = new URL(url);
      const transport = parsedUrl.protocol === 'https:' ? https : http;

      const options: http.RequestOptions = {
        method: 'POST',
        hostname: parsedUrl.hostname,
        port: parsedUrl.port || (parsedUrl.protocol === 'https:' ? 443 : 80),
        path: parsedUrl.pathname + parsedUrl.search,
        headers: {
          ...headers,
          'Content-Length': Buffer.byteLength(body),
        },
        timeout: this._config.timeout * 1000,
      };

      const req = transport.request(options, (res) => {
        const chunks: Buffer[] = [];
        res.on('data', (chunk: Buffer) => chunks.push(chunk));
        res.on('end', () => {
          const statusCode = res.statusCode || 0;
          if (statusCode >= 200 && statusCode < 300) {
            resolve(Buffer.concat(chunks));
          } else {
            const responseBody = Buffer.concat(chunks).toString('utf-8');
            const responseHeaders: Record<string, string> = {};
            for (const [key, value] of Object.entries(res.headers)) {
              if (typeof value === 'string') {
                responseHeaders[key.toLowerCase()] = value;
              }
            }
            reject(new HttpError(statusCode, responseBody, responseHeaders));
          }
        });
      });

      req.on('error', (err) => reject(err));
      req.on('timeout', () => {
        req.destroy();
        reject(new Error(`Request timed out after ${this._config.timeout}s`));
      });

      req.write(body);
      req.end();
    });
  }

  protected _httpGetJson(
    url: string,
    headers: Record<string, string>
  ): Promise<Record<string, any>> {
    return new Promise((resolve, reject) => {
      const parsedUrl = new URL(url);
      const transport = parsedUrl.protocol === 'https:' ? https : http;

      const options: http.RequestOptions = {
        method: 'GET',
        hostname: parsedUrl.hostname,
        port: parsedUrl.port || (parsedUrl.protocol === 'https:' ? 443 : 80),
        path: parsedUrl.pathname + parsedUrl.search,
        headers,
        timeout: this._config.timeout * 1000,
      };

      const req = transport.request(options, (res) => {
        const chunks: Buffer[] = [];
        res.on('data', (chunk: Buffer) => chunks.push(chunk));
        res.on('end', () => {
          const responseBody = Buffer.concat(chunks).toString('utf-8');
          const statusCode = res.statusCode || 0;
          if (statusCode >= 200 && statusCode < 300) {
            try {
              resolve(JSON.parse(responseBody));
            } catch {
              reject(new Error(`Invalid JSON response: ${responseBody.substring(0, 200)}`));
            }
          } else {
            const responseHeaders: Record<string, string> = {};
            for (const [key, value] of Object.entries(res.headers)) {
              if (typeof value === 'string') {
                responseHeaders[key.toLowerCase()] = value;
              }
            }
            reject(new HttpError(statusCode, responseBody, responseHeaders));
          }
        });
      });

      req.on('error', (err) => reject(err));
      req.on('timeout', () => {
        req.destroy();
        reject(new Error(`Request timed out after ${this._config.timeout}s`));
      });

      req.end();
    });
  }

  protected _httpGetText(
    url: string,
    headers: Record<string, string>
  ): Promise<string | Record<string, any>> {
    return new Promise((resolve, reject) => {
      const parsedUrl = new URL(url);
      const transport = parsedUrl.protocol === 'https:' ? https : http;

      const options: http.RequestOptions = {
        method: 'GET',
        hostname: parsedUrl.hostname,
        port: parsedUrl.port || (parsedUrl.protocol === 'https:' ? 443 : 80),
        path: parsedUrl.pathname + parsedUrl.search,
        headers,
        timeout: this._config.timeout * 1000,
      };

      const req = transport.request(options, (res) => {
        const chunks: Buffer[] = [];
        res.on('data', (chunk: Buffer) => chunks.push(chunk));
        res.on('end', () => {
          const responseBody = Buffer.concat(chunks).toString('utf-8');
          const statusCode = res.statusCode || 0;
          if (statusCode >= 200 && statusCode < 300) {
            try {
              resolve(JSON.parse(responseBody));
            } catch {
              resolve(responseBody);
            }
          } else {
            const responseHeaders: Record<string, string> = {};
            for (const [key, value] of Object.entries(res.headers)) {
              if (typeof value === 'string') {
                responseHeaders[key.toLowerCase()] = value;
              }
            }
            reject(new HttpError(statusCode, responseBody, responseHeaders));
          }
        });
      });

      req.on('error', (err) => reject(err));
      req.on('timeout', () => {
        req.destroy();
        reject(new Error(`Request timed out after ${this._config.timeout}s`));
      });

      req.end();
    });
  }

  protected _sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: src/python/modelmesh/cdk/base_provider.py
"""Base provider connector implementation.

Implements the full ``ProviderConnector`` interface with OpenAI-compatible
default behavior.  Subclasses override protected hook methods to adapt
to non-OpenAI APIs without reimplementing transport, retries, or error
classification.

HTTP transport uses :mod:`urllib.request` from the standard library so
the package has zero external dependencies.
"""
from __future__ import annotations

import asyncio
import json
import urllib.error
import urllib.request
from dataclasses import dataclass, field
from typing import AsyncIterator, Optional

from modelmesh.interfaces.provider import (
    ChatMessage,
    CompletionChoice,
    CompletionRequest,
    CompletionResponse,
    ErrorClassification,
    ModelInfo,
    ModelPricing,
    ProviderConnector,
    QuotaStatus,
    RateLimitStatus,
    TokenUsage,
)

__all__ = [
    "BaseProviderConfig",
    "BaseProvider",
]


@dataclass
class BaseProviderConfig:
    """Configuration for a BaseProvider instance."""

    base_url: str = ""
    api_key: str = ""
    models: list[ModelInfo] = field(default_factory=list)
    timeout: float = 30.0
    max_retries: int = 3
    auth_method: str = "api_key"
    retryable_codes: list[int] = field(
        default_factory=lambda: [429, 500, 502, 503]
    )
    non_retryable_codes: list[int] = field(
        default_factory=lambda: [400, 401, 403]
    )
    capabilities: list[str] = field(
        default_factory=lambda: ["generation.text-generation.chat-completion"]
    )


class BaseProvider(ProviderConnector):
    """Base implementation of the ProviderConnector interface.

    Provides an OpenAI-compatible default behavior for all methods.
    Subclasses override protected hook methods to adapt to non-OpenAI
    APIs without reimplementing transport, retries, or error handling.

    HTTP transport uses :mod:`urllib.request` from the standard library.
    Async methods delegate blocking I/O to a thread via
    :func:`asyncio.to_thread`.
    """

    def __init__(
        self, config: BaseProviderConfig, observability=None
    ) -> None:
        self._config = config
        self._request_count: int = 0
        self._tokens_used: int = 0
        self._models_by_id: dict[str, ModelInfo] = {
            m.id: m for m in config.models
        }
        self._observability = observability

    def _trace(
        self,
        severity,
        component: str,
        message: str,
        error: str | None = None,
        **metadata,
    ) -> None:
        """Emit a trace entry through the observability connector."""
        from datetime import datetime

        from modelmesh.interfaces.observability import Severity as SevEnum
        from modelmesh.interfaces.observability import TraceEntry

        sev = severity if isinstance(severity, SevEnum) else SevEnum(severity.lower())
        entry = TraceEntry(
            severity=sev,
            timestamp=datetime.now(),
            component=component,
            message=message,
            error=error,
            metadata=metadata if metadata else None,
        )
        if self._observability:
            self._observability.trace(entry)

    # ── Model Execution ─────────────────────────────────────────────

    async def complete(self, request: CompletionRequest) -> CompletionResponse:
        """Send a completion request and return the full response.

        Builds the payload via ``_build_request_payload``, posts to the
        endpoint returned by ``_get_completion_endpoint``, and parses
        the response via ``_parse_response``.  Retries on retryable
        status codes up to ``max_retries`` times.
        """
        import time as _time

        component = f"provider.{request.model}"
        self._trace(
            "DEBUG",
            component,
            f"Sending completion request for model '{request.model}'",
            model=request.model,
        )

        payload = self._build_request_payload(request)
        headers = self._build_headers()
        endpoint = self._get_completion_endpoint()
        start_time = _time.monotonic()

        last_error: Optional[Exception] = None
        for attempt in range(self._config.max_retries + 1):
            try:
                data = await asyncio.to_thread(
                    self._http_post, endpoint, payload, headers
                )
                result = self._parse_response(data)
                self.report_usage(request.model, result.usage)
                latency_ms = (_time.monotonic() - start_time) * 1000
                self._trace(
                    "INFO",
                    component,
                    f"Completion succeeded for model '{request.model}'",
                    model=request.model,
                    latency_ms=round(latency_ms, 2),
                    prompt_tokens=result.usage.prompt_tokens,
                    completion_tokens=result.usage.completion_tokens,
                    total_tokens=result.usage.total_tokens,
                )
                return result
            except Exception as exc:
                last_error = exc
                classification = self.classify_error(exc)
                if (
                    not classification.retryable
                    or attempt == self._config.max_retries
                ):
                    self._trace(
                        "ERROR",
                        component,
                        f"Non-retryable error for model "
                        f"'{request.model}': {exc}",
                        error=str(exc),
                        model=request.model,
                        attempt=attempt + 1,
                        category=classification.category,
                    )
                    raise
                self._trace(
                    "WARNING",
                    component,
                    f"Retryable error for model '{request.model}' "
                    f"(attempt {attempt + 1}): {exc}",
                    error=str(exc),
                    model=request.model,
                    attempt=attempt + 1,
                    category=classification.category,
                )
                retry_after: float = 2 ** attempt
                # Check for Retry-After header on urllib HTTPError
                raw_retry = getattr(
                    getattr(exc, "headers", None), "get", lambda _: None
                )("Retry-After")
                if raw_retry is not None:
                    try:
                        retry_after = float(raw_retry)
                    except (ValueError, TypeError):
                        pass
                await asyncio.sleep(retry_after)

        raise last_error  # type: ignore[misc]

    async def stream(
        self, request: CompletionRequest
    ) -> AsyncIterator[CompletionResponse]:
        """Send a completion request and yield partial responses via SSE.

        Sets ``stream: true`` in the payload and reads the response as
        a server-sent event stream, parsing each chunk through
        ``_parse_sse_chunk``.
        """
        self._trace(
            "DEBUG",
            f"provider.{request.model}",
            f"Starting streaming request for model '{request.model}'",
            model=request.model,
        )
        payload = self._build_request_payload(request)
        payload["stream"] = True
        headers = self._build_headers()
        endpoint = self._get_completion_endpoint()

        lines = await asyncio.to_thread(
            self._http_post_stream, endpoint, payload, headers
        )
        for line in lines:
            if not line or not line.startswith("data: "):
                continue
            data_str = line[len("data: "):]
            if data_str.strip() == "[DONE]":
                break
            chunk = self._parse_sse_chunk(data_str)
            if chunk is not None:
                yield chunk

    # ── Capabilities ────────────────────────────────────────────────

    def get_capabilities(self) -> list[str]:
        """Return the list of capability identifiers this provider supports."""
        return list(self._config.capabilities)

    def supports(self, capability: str) -> bool:
        """Check whether this provider supports a specific capability."""
        return capability in self._config.capabilities

    # ── Model Catalogue ─────────────────────────────────────────────

    def list_models(self) -> list[ModelInfo]:
        """Return all models available from this provider."""
        return list(self._config.models)

    def get_model_info(self, model_id: str) -> ModelInfo:
        """Return detailed information for a specific model.

        Raises:
            KeyError: If the model ID is not found in the catalogue.
        """
        if model_id not in self._models_by_id:
            raise KeyError(f"Model not found: {model_id}")
        return self._models_by_id[model_id]

    # ── Quota & Rate Limits ─────────────────────────────────────────

    def check_quota(self) -> QuotaStatus:
        """Return current quota consumption.  No limit enforced by default."""
        return QuotaStatus(used=self._request_count)

    def get_rate_limits(self) -> RateLimitStatus:
        """Return current rate-limit headroom.  Unknown by default."""
        return RateLimitStatus()

    # ── Cost & Pricing ──────────────────────────────────────────────

    def get_pricing(self, model_id: str) -> ModelPricing:
        """Return pricing information for a specific model.

        Raises:
            KeyError: If the model or its pricing is not configured.
        """
        info = self.get_model_info(model_id)
        if info.pricing is None:
            raise KeyError(f"No pricing configured for model: {model_id}")
        return info.pricing

    def report_usage(self, model_id: str, usage: TokenUsage) -> None:
        """Increment internal request and token counters."""
        self._request_count += 1
        self._tokens_used += usage.total_tokens

    # ── Error Classification ────────────────────────────────────────

    def classify_error(self, error: Exception) -> ErrorClassification:
        """Classify an error using configured retryable/non-retryable codes.

        Maps HTTP status codes to categories: ``rate_limit`` for 429,
        ``server`` for 5xx, ``client`` for 4xx, and ``unknown`` for
        everything else.
        """
        # Support both urllib.error.HTTPError (.code) and httpx-style
        # (.response.status_code) error objects.
        status_code: Optional[int] = getattr(error, "code", None)
        if status_code is None:
            status_code = getattr(
                getattr(error, "response", None), "status_code", None
            )
        if status_code is None:
            result = ErrorClassification(retryable=False, category="unknown")
            self._trace(
                "DEBUG",
                "provider",
                f"Error classified as '{result.category}' "
                f"(retryable={result.retryable})",
                error=str(error),
                category=result.category,
                retryable=result.retryable,
            )
            return result

        if status_code in self._config.retryable_codes:
            result = ErrorClassification(
                retryable=True,
                error_code=status_code,
                category="rate_limit" if status_code == 429 else "server",
            )
            self._trace(
                "DEBUG",
                "provider",
                f"Error classified as '{result.category}' "
                f"(retryable={result.retryable}, code={status_code})",
                error=str(error),
                status_code=status_code,
                category=result.category,
                retryable=result.retryable,
            )
            return result

        if status_code in self._config.non_retryable_codes:
            category = "auth" if status_code in (401, 403) else "client"
            result = ErrorClassification(
                retryable=False,
                error_code=status_code,
                category=category,
            )
            self._trace(
                "DEBUG",
                "provider",
                f"Error classified as '{result.category}' "
                f"(retryable={result.retryable}, code={status_code})",
                error=str(error),
                status_code=status_code,
                category=result.category,
                retryable=result.retryable,
            )
            return result

        result = ErrorClassification(
            retryable=False, error_code=status_code, category="unknown"
        )
        self._trace(
            "DEBUG",
            "provider",
            f"Error classified as '{result.category}' "
            f"(retryable={result.retryable}, code={status_code})",
            error=str(error),
            status_code=status_code,
            category=result.category,
            retryable=result.retryable,
        )
        return result

    def is_retryable(self, error: Exception) -> bool:
        """Return True if the error is eligible for retry."""
        return self.classify_error(error).retryable

    # ── Protected Hooks ─────────────────────────────────────────────

    def _build_request_payload(self, request: CompletionRequest) -> dict:
        """Translate a CompletionRequest into an OpenAI-format JSON payload.

        Override this method to adapt the request format for
        non-OpenAI-compatible APIs.
        """
        payload: dict = {
            "model": request.model,
            "messages": request.messages,
        }
        if request.temperature is not None:
            payload["temperature"] = request.temperature
        if request.max_tokens is not None:
            payload["max_tokens"] = request.max_tokens
        if request.tools:
            payload["tools"] = request.tools
        if request.stream:
            payload["stream"] = True
        return payload

    def _parse_response(self, data: dict) -> CompletionResponse:
        """Parse an OpenAI-format JSON response into a CompletionResponse.

        Override this method to handle custom response schemas.
        """
        usage_data = data.get("usage", {})
        raw_choices = data.get("choices", [])
        choices: list[CompletionChoice] = []
        for raw in raw_choices:
            msg_data = raw.get("message")
            message = None
            if msg_data:
                message = ChatMessage(
                    role=msg_data.get("role", "assistant"),
                    content=msg_data.get("content"),
                    tool_calls=msg_data.get("tool_calls"),
                )
            choices.append(
                CompletionChoice(
                    index=raw.get("index", 0),
                    message=message,
                    finish_reason=raw.get("finish_reason"),
                )
            )
        return CompletionResponse(
            id=data.get("id", ""),
            model=data.get("model", ""),
            choices=choices,
            usage=TokenUsage(
                prompt_tokens=usage_data.get("prompt_tokens", 0),
                completion_tokens=usage_data.get("completion_tokens", 0),
                total_tokens=usage_data.get("total_tokens", 0),
            ),
        )

    def _build_headers(self) -> dict[str, str]:
        """Build HTTP headers for the request.

        Override to add custom headers or change the auth scheme.
        """
        headers: dict[str, str] = {"Content-Type": "application/json"}
        if self._config.api_key:
            headers["Authorization"] = f"Bearer {self._config.api_key}"
        return headers

    def _get_completion_endpoint(self) -> str:
        """Return the full URL for the chat completions endpoint.

        Override to change the URL path for non-OpenAI APIs.
        """
        base = self._config.base_url.rstrip("/")
        return f"{base}/v1/chat/completions"

    def _parse_sse_chunk(self, line: str) -> CompletionResponse | None:
        """Parse a single SSE data line into a partial CompletionResponse.

        Override to handle non-standard streaming formats.  Return
        ``None`` to skip a chunk.
        """
        try:
            data = json.loads(line)
        except json.JSONDecodeError:
            return None
        raw_choices = data.get("choices", [])
        if not raw_choices:
            return None
        choices: list[CompletionChoice] = []
        for raw in raw_choices:
            delta_data = raw.get("delta")
            delta = None
            if delta_data:
                delta = ChatMessage(
                    role=delta_data.get("role", "assistant"),
                    content=delta_data.get("content"),
                    tool_calls=delta_data.get("tool_calls"),
                )
            choices.append(
                CompletionChoice(
                    index=raw.get("index", 0),
                    delta=delta,
                    finish_reason=raw.get("finish_reason"),
                )
            )
        return CompletionResponse(
            id=data.get("id", ""),
            model=data.get("model", ""),
            choices=choices,
            usage=TokenUsage(
                prompt_tokens=0, completion_tokens=0, total_tokens=0
            ),
        )

    async def close(self) -> None:
        """Release any resources held by this connector."""
        pass

    # ── HTTP Transport (stdlib) ─────────────────────────────────────

    def _http_post(
        self, url: str, payload: dict, headers: dict[str, str]
    ) -> dict:
        """Send a synchronous HTTP POST and return the parsed JSON body.

        Uses :mod:`urllib.request` so the package has zero external
        dependencies.  Called from async code via
        :func:`asyncio.to_thread`.
        """
        body = json.dumps(payload).encode("utf-8")
        req = urllib.request.Request(
            url, data=body, headers=headers, method="POST"
        )
        with urllib.request.urlopen(req, timeout=self._config.timeout) as resp:
            return json.loads(resp.read().decode("utf-8"))

    def _http_post_stream(
        self, url: str, payload: dict, headers: dict[str, str]
    ) -> list[str]:
        """Send a synchronous HTTP POST and return raw SSE lines.

        Uses :mod:`urllib.request` so the package has zero external
        dependencies.  Called from async code via
        :func:`asyncio.to_thread`.
        """
        body = json.dumps(payload).encode("utf-8")
        req = urllib.request.Request(
            url, data=body, headers=headers, method="POST"
        )
        with urllib.request.urlopen(req, timeout=self._config.timeout) as resp:
            raw = resp.read().decode("utf-8")
        return raw.splitlines()
